src/ui.rs

```diff
@@ -183,3 +220,85 @@ fn centered_rect(percent_x: u16, percent_y: u16, r: Rect) -> Rect {
 
     center
 }
+
+pub fn tui_loop(
+    cpu: &mut CPU,
+    uart_rx: &mpsc::Receiver<char>,
+    uart_tx: &mpsc::Sender<char>,
+) -> anyhow::Result<()> {
+    enable_raw_mode()?;
+    let stdout = io::stdout();
+    let backend = CrosstermBackend::new(stdout);
+    let mut terminal = Terminal::new(backend)?;
+    let _ = terminal.clear();
+
+    let mut ui = ViewState::new();
+    let mut show_help = true;
+    let mut user_input = String::new();
+    let mut insert_mode = false;
+    let mut auto_step = false;
+
+    loop {
+        terminal.draw(|f| ui.ui(f, cpu, uart_rx, show_help, insert_mode, &user_input))?;
+
+        if let Ok(maybe_event) = event::poll(Duration::from_millis(50)) {
+            if maybe_event {
+                if let Event::Key(key) = event::read()? {
+                    if insert_mode {
```

Comment:
@edjopato I have a lot of nesting here. Do you think there is a more elegant way of doing the `poll & query & matching`?

Reply:
(and please don't invest too much time here, this is likely to get refactored anyways soonish)

Reply:
Maybe with [`Result::is_ok_and`](https://doc.rust-lang.org/std/result/enum.Result.html#method.is_ok_and)? Or just `poll(…)?` / `poll(…).expect(…)`? Event handling is a core feature required, when it doesn't work the application is kinda pointless? Then its just `if event::poll(…)? {…}`.

The `if insert_mode` seems needed for the logic.

When event handling starts to involve something else than keys the `event::read()` can be better handled by a match directly. (`if event::poll(…)? { match event::read()? { Event::Key(KeyCode::Enter)) => {…} } }`. Not sure whether thats a benefit in your case…

Reply:
Thanks 💛!
You hint towards `if event::poll(…)? {…}` did the trick!